heclib/hecdss/hecdss.c

```diff
@@ -0,0 +1,742 @@
+
+    zStructTransfer* transfer = zstructTransferNew(pathname, 0);
+    zStructTimeSeries* tss = zstructTsNew(pathname);
+    transfer->internalHeaderMode = 1;
+    int status = zread(pdss->ifltab, transfer);
+    if (status == 0)
+    {
+      int intervalType = ztsProcessTimes(pdss->ifltab, tss, 0);
+      status = ztsInternalHeaderUnpack(tss, transfer->internalHeader, transfer->internalHeaderNumber);
+      stringCopy(units, unitsLength, tss->units, strlen(tss->units));
+      stringCopy(type, typeLength, tss->type, strlen(tss->type));
+    }
+    zstructFree(tss);
+    zstructFree(transfer);
+    return status;
+}
+
+
+HECDSS_API int hec_dss_tsRetrieve(dss_file* dss, const char *pathname, 
+                                  const char *startDate, const char *startTime, 
+                                  const char* endDate,   const char *endTime,
+                                  int *timeArray, double *valueArray, const int arraySize,
+                                  int *numberValuesRead, int *quality, int* qualityLength,
+                                  int* julianBaseDate,int* timeGranularitySeconds,
+                                  char* units,const int unitsLength, char* type,const int typeLength)
+{
+    *numberValuesRead = 0; 
+    zStructTimeSeries* tss = zstructTsNew(pathname);
+
+    tss->startJulianDate = dateToJulian(startDate);
+    tss->startTimeSeconds = timeStringToSeconds(startTime);
+    tss->endJulianDate = dateToJulian(endDate);
+    tss->endTimeSeconds = timeStringToSeconds(endTime);
+    
+    const int retrieveDoublesFlag = 2; // get doubles
+    const int boolRetrieveAnyQualityNotes = 1;
+    /*int retrieveFlag
+      * A flag indicating how to readand trim data.
+      * For regular interval data :
+    *retrieveFlag = 0 : Adhere to time window provided and generate the time array.
+      * retrieveFlag = -1 : Trim data.Remove missing values at the beginning and end of data set(not inside),
+      *andgenerate the time array.
+      * retrieveFlag = -2 : Adhere to time window provided but do not include time array.
+      * retrieveFlag = -3 : Trim data.Remove missing values at the beginning and end of data set(not inside),
+      * no time array.
+      * For irregular interval data :
+    *retrieveFlag = 0 : Adhere to time window provided.
+      * retrieveFlag = 1 : Retrieve one value previous to start of time window
+      * retrieveFlag = 2 : Retrieve one value after end of time window
+      * retrieveFlag = 3 : Retrieve one value before and one value after time window
+      */
+    const int retrieveUsingTimeWindowFlag = 0; // Adhere to time window provided and generate the time array
+
+    int status = ztsRetrieve(dss->ifltab, tss, retrieveUsingTimeWindowFlag,retrieveDoublesFlag, boolRetrieveAnyQualityNotes);
+    if (status == 0) {
+        *julianBaseDate = tss->julianBaseDate;
+        *timeGranularitySeconds = tss->timeGranularitySeconds;
+        
```

Comment:
Initialize char* out parameters below to empty strings either before `stringCopy(...)` or in `else` clauses.